SUMO-77267 Add support for configuring SourceHost and SourceCategory in .NET log appenders

yuting-liu opened the pull request:
###### Description
Add support for configuring SourceHost and SourceCategory in .NET log appenders

###### Testing performed
- [X] Ran unit tests in modified modules ( `.\SumoLogic.Logging.CI.ps1` )

**To be merged by:** @Yuting
**Delete branch after merge:** Yes
**Track deployment for:** all

TerribleDev commented on SumoLogic.Logging.Common/Sender/SumoLogicMessageSender.cs:
```diff
@@ -182,7 +212,17 @@ public void TrySend(string body, string name)
+                httpContent.Headers.Add(SUMO_SOURCE_NAME_HEADER, name);
+                // set headers for the source category and the source host
+                // when they are not null nor empty strings
+                if(SourceCategory != null || !SourceCategory.Trim().Equals(""))
```
You can replace this with: 

```csharp 

if(!String.IsNullOrWhitespace(sourceCategory))
{
 ...
}

```

TerribleDev commented on SumoLogic.Logging.Common/Sender/SumoLogicMessageSender.cs:
```diff
@@ -182,7 +212,17 @@ public void TrySend(string body, string name)
+                {
+                    httpContent.Headers.Add(SUMO_SOURCE_CATEGORY_HEADER, SourceCategory);
+                }
+                if (SourceHost != null || !SourceHost.Trim().Equals(""))
```
Same as above, you can use `!String.IsNullOrWhitespace(sourceHost)`

TerribleDev commented on SumoLogic.Logging.NLog/BufferedSumoLogicTarget.cs:
```diff
@@ -254,6 +272,8 @@ protected override void InitializeTarget()
             this.SumoLogicMessageSender.RetryInterval = TimeSpan.FromMilliseconds(this.RetryInterval);
             this.SumoLogicMessageSender.ConnectionTimeout = TimeSpan.FromMilliseconds(this.ConnectionTimeout);
             this.SumoLogicMessageSender.Url = string.IsNullOrEmpty(this.Url) ? null : new Uri(this.Url);
+            this.SumoLogicMessageSender.SourceCategory = string.IsNullOrEmpty(this.SourceCategory) ? null : this.SourceCategory;
```
You should probably guard against whitespace as well. `string.IsNullOrWhitespace`

yuting-liu replied:
It's great to know this function `isNullOrWhiteSpace`! Already fix it.

latkin commented on SumoLogic.Logging.Common/Sender/SumoLogicMessageSender.cs:
```diff
@@ -182,7 +203,17 @@ public void TrySend(string body, string name)
 
             using (var httpContent = new StringContent(body, Encoding.UTF8, TextPlainMediaType))
             {
-                httpContent.Headers.Add("X-Sumo-Name", name);
```
I would like to keep the model consistent -- instead of setting category and host via properties, but passing name as an arugment, let's just make everything a property.

latkin commented on SumoLogic.Logging.Common.Tests/Sender/SumoLogicMessageSenderTest.cs:
```diff
@@ -69,7 +76,7 @@ public SumoLogicMessageSenderTest()
         {
             Assert.Equal(0, this.messagesHandler.ReceivedRequests.Count);
-            this.sumoLogicMessageSender.Send("name", "body");
+            this.sumoLogicMessageSender.Send("name");
```
I think you want to pass `"body"` and remove `"name"`

latkin replied:
Or rather, looks like the code previously had a minor bug where it was using `"name"` and the body, and `"body"` as the name

latkin commented on SumoLogic.Logging.Common.Tests/Sender/SumoLogicMessageSenderTest.cs:
```diff
@@ -118,7 +125,7 @@ public void HttpResponseCodeTest()
             Assert.True(this.sumoLogicMessageSender.CanSend);
             Assert.True(this.sumoLogicMessageSender.CanTrySend);
-            this.sumoLogicMessageSender.Send("name", "body");
+            this.sumoLogicMessageSender.Send("name");
```
Same minor bug here

yuting-liu replied:
Just fixed them : )

latkin commented on SumoLogic.Logging.Common/Sender/SumoLogicMessageSender.cs:
```diff
@@ -123,14 +153,14 @@ private HttpClient HttpClient
         /// <param name="name">The message name.</param>
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
-        public void Send(string body, string name)
+        public void Send(string body)
```
Just realized that this is a public API, so we can't change its signature. I will discuss solution with you

latkin commented on SumoLogic.Logging.Common/Sender/SumoLogicMessageSender.cs:
```diff
@@ -164,6 +193,16 @@ public void Send(string body, string name)
+        /// Blocks while sending a message to the SumoLogic server, retrying as many time as needed.
+        /// </summary>
+        /// <param name="body">The message body.</param>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
```
The `SuppressMessage` attribute should stay on the `Send(body, name)` method

latkin commented on SumoLogic.Logging.Common/Sender/SumoLogicMessageSender.cs:
```diff
@@ -64,6 +67,33 @@ public Uri Url
         }
 
         /// <summary>
+        /// Gets or ses the source name
```
typo - `ses` 😄

latkin commented on SumoLogic.Logging.Common/Sender/SumoLogicMessageSender.cs:
```diff
@@ -164,6 +194,16 @@ public void Send(string body, string name)
+        /// Blocks while sending a message to the SumoLogic server, retrying as many time as needed.
+        /// </summary>
+        /// <param name="body">The message body.</param>
+        [Obsolete("Set the SourceName property and use Send(string body)")]
```
`Obsolete` goes on the OLD APIs with (body, name) signature, not the new ones

rvmiller89 commented on SumoLogic.Logging.Common/Sender/SumoLogicMessageSender.cs:
```diff
@@ -182,7 +200,18 @@ public void TrySend(string body, string name)
             using (var httpContent = new StringContent(body, Encoding.UTF8, TextPlainMediaType))
             {
-                httpContent.Headers.Add("X-Sumo-Name", name);
+                if (!string.IsNullOrWhiteSpace(name))
```
`if (!string` should be `if (!String` ?

yuting-liu replied:
I googled the difference between `string` and `String` in C#. It says`string` is an alias for the `String` class in .NET framework. So technically there is no difference, but I'm wondering whether we have code style rules for this?

rvmiller89 replied:
why not be consistent with 2 usages below though?  Or change them all to `string`?

yuting-liu replied:
Okay, I would change it to all `String`.

latkin replied:
the keyword `string` is an alias to the class `System.String` so it doesn't make a difference, besides style

rvmiller89 commented:
should we update the README with usage and examples in the same PR?

latkin commented:
@rvmiller89 yeah might has well. @yuting-liu can you update the README to show the new fields in the configuration XML

yuting-liu commented:
I changed `README.md` with the examples of the configuration XML. And also the descriptions of appenders in `README.txt` in `Log4Net` and `NLog` folders. BTW, `README.txt` was in `Log4Net.Example` before, which I moved into `Log4Net` now.

latkin commented:
LGTM

TerribleDev commented:
👍
